integration/doc/oasis/charge_station_based_routing.md

```diff
@@ -0,0 +1,88 @@
+
+![](../graph/charge_station_based_routing.png)
+
+If we pro-process all charge stations in the graph and record the `connectivity` for each charge station, which means for each single charge station, we could retrieve  **all reachable charge stations sorted by distance or energy needed**, then we could calculate optimal result based on classic algorithm.
+
+
+## Design
+
+### Pre-process work flow
+
+```
+Telenav's Database(postGIS) 
+       -> csv 
+               -> adjust column name 
+                       -> convert csv to json 
+                                 -> decoding json
+```
+More information [#237](https://github.com/Telenav/osrm-backend/issues/237) 
+
+
+
+### JSON Input
+Electricity charge stations, expect input in JSON format follow OSM's tag definition about [node](https://wiki.openstreetmap.org/wiki/Node) and [charging_station](https://wiki.openstreetmap.org/wiki/Tag:amenity%3Dcharging_station), example:
+```json
+{
+   "id": 12345,
+   "lat": 51.5173639,
+   "lon": -0.140043,
+   "amenity": "charging_station",
+   "operator": "NOFT",
+   "fee": "yes",
+   "capacity":4,
+}
+```
+- Later could implement more input connector: pbf, csv, etc.
+
+### Pre-Processing
+- For each charge station point, calculate its google s2 cellid
+   + For each point, it could have cellid in [0, 30] levels.  We just pick a range from them, such as [10, 20].
+   + The range used for pre-processing will be the same as query
+- For each charge station point, take which as center and construct a circle whose radius is a hard code max drive range value(500km), query for all cells it could touch, only record the one contains charge stations
+  + Make sure such function could work: give a cell id(not station id) could retrieve all cells it could reached by.  We also need to distinguish the stations could be reached by charging how much of energy
+  + Make sure exact distance between charge stations has been recorded in db
+  + For production, circle should be replaced by `reach range`(or `isoline`, consider elevation, red/lights, energy pattern, etc)
+- More infor about google S2
+  + [How S2 cell is generated](https://github.com/CodeBear801/tech_summary/blob/master/tech-summary/navigation/spatial_index/google_s2_cell.md)
+  + [How hilbert curve works](https://github.com/CodeBear801/tech_summary/blob/master/tech-summary/navigation/spatial_index/hilbert_curve.md)
+  + [#236](https://github.com/Telenav/osrm-backend/issues/236)
+
+### Query
+- Given two points of orig/dest
+- Check whether orig/dest is reachable by current energy
+- If not start two strategy together: `charge station based routing` and `search along route`
+- logic of `charge station routing`
+   + For input, need using its current energy to filter all charge station it could reached
+   + For Dest, check all charge station needed to reach destination
+        *  Check whether there is overlap between those two, if yes, then go to next step
```

Comment:
Question: if no overlap, currently the algorithm will be ignored, right? On the other hand, we at least already got some mandatory stations. Find shortest path on the station graph with given orig/dest and these marked mandatory stations may also possible, how's your opinion?

Reply:
It's possible.  I think the idea logic is:
- generate an iterator represent all reachable charge stations from orig, which represent `mandatory stations`
- generate an iterator represent all reachable charge stations from dest, which represent `mandatory stations`
- if no overlap, those iterators will be used in graph for next step

`Iterators` has been implemented but I didn't used it in graph calculation yet.